### PR TITLE
Exponential backoff for recovery retry interval

### PR DESCRIPTION
**Pull request by vikinghawk**
Isssue #308 

First stab at a pluggable recovery retry interval and basic fibonnaci backoff implementation. Let me know your thoughts and i can make changes and add javadoc if this seems like something you guys like.

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@vikinghawk Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-java-client&pullRequestId=309)!

[Click here](https://cla.pivotal.io/sync/rabbitmq?repositoryId=rabbitmq/rabbitmq-java-client&pullRequestId=309) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@vikinghawk Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/rabbitmq?repositoryId=rabbitmq/rabbitmq-java-client&pullRequestId=309)!

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java [hunk 1]
```diff
@@ -67,21 +67,26 @@ public void close() throws IOException, TimeoutException {
...
-          this.connection.unregisterChannel(this);
+            recoveryCleanup();
         }
     }
+    
+    private void recoveryCleanup() {
+        for (String consumerTag : consumerTags) {
+            this.connection.deleteRecordedConsumer(consumerTag);
+        }
+        // TODO what about any other recorded queues and bindings that were owned by this channel? They will now cause recovery exceptions.
```
Comment (michaelklishin) [inline src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java]: Queues and bindings are not "owned" by channels in any way. Consumers are channel state, however. Queues and bindings are tracked by connections (in part because queues can be owned by connections — when they are exclusive — and in part because declaring something on one channel then deleting it on another should leave no traces of that entity.

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java [hunk 2]
```diff
@@ -67,21 +67,26 @@ public void close() throws IOException, TimeoutException {
         try {
             delegate.close();
         } finally {
-            for (String consumerTag : consumerTags) {
-                this.connection.deleteRecordedConsumer(consumerTag);
-            }
-            this.connection.unregisterChannel(this);
+            recoveryCleanup();
```
Comment (michaelklishin) [inline src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java]: Is this change really related to delay calculation and back-off?

**Comment by michaelklishin** [comment]
In general this looks promising.

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java (hunk 2 above)
Comment (vikinghawk) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java]: no, just noticed the recorded consumers weren't being deleted with the 2nd close method like the 1st.

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java (hunk 1 above)
Comment (vikinghawk) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java]: Ya thats a bad description. What that is referencing is that any queues or bindings that were created with this channel still have the recorded entities reference this now closed channel and will throw an exception during recovery. I don't know a good way to address this other than having recorded queues/exchanges/bindings use a special "recovery" channel instead of the original creating channel. I assume the only recorded items that would require to be tied to the original channel that created them would be consumers which are cleaned up.

Comment (vikinghawk) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java]: removed this change.

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java (hunk 2 above)
Comment (vikinghawk) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringChannel.java]: removed for now. could be a separate future PR

**Comment by vikinghawk** [comment]
Updated with javadoc, test, and suggested changes.

File: src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java [hunk 3]
```diff
@@ -541,8 +542,7 @@ private RecoveryAwareAMQConnection recoverConnection() throws InterruptedExcepti
 				newConn.abort();
 				return null;
             } catch (Exception e) {
-                // TODO: exponential back-off
-                Thread.sleep(this.params.getNetworkRecoveryInterval());
+                Thread.sleep(this.params.getRecoveryDelayHandler().getDelay(attempts));
```
Comment (vikinghawk) [inline src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java]: thoughts on flipping the order of this? so deliver the exception before sleeping?

Comment (acogoluegnes) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java]: Let's keep it the way it is.

Comment (michaelklishin) [reply src/main/java/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java]: Good catch. I think at some point we will have to do it. The way it works right now is a bit weird and can complicate additional manual recovery steps.
